build: add workflow and script to leave comment with make instructions

Planeshifter opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   adds slash command for posting a PR-specific command comment with relevant `make` command.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/9.

## Questions

> Any questions for reviewers of this pull request?

When multiple packages are modified in one PR, the command currently does not leave any comment. Maybe that's not desirable. We could post a generic comment of the relevant commands where the expectation would then be that contributors themselves have to set the relevant environment variables. Thoughts?

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Planeshifter committed:
build: add workflow and script to leave comment with make instructions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: missing_dependencies
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented on .github/workflows/slash_commands.yml:
```diff
@@ -218,6 +236,7 @@ jobs:
             @${{ github.event.comment.user.login }}, available slash commands include:
 
             - `/stdlib check-files` - Check for required files.
+            - `/stdlib commands-comment` - Print `make` commands for package changed in PR.
```
@Planeshifter Can we rename this to `make-commands`? The `-comment` suffix seems to leak an implementation detail.

Planeshifter replied:
Done.

Planeshifter committed:
build: rename slash command

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented:
> the command currently does not leave any comment. Maybe that's not desirable. We could post a generic comment of the relevant commands where the expectation would then be that contributors themselves have to set the relevant environment variables. Thoughts?

Yes, we probably need to leave some sort of comment; otherwise, devs are going to wonder what's wrong and why the command didn't do anything. Devs aren't going to be aware as to why the comment never posts and they may just wait around hoping that one eventually shows up.

Planeshifter committed:
build: simplify and leave comment when multiple packages are changed

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: passed
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter commented:
@kgryte Now leaving the following comment when multiple packages are changed:

====

Hello! 👋

Pro-tip: This PR changes multiple packages. You can use various `make` rules with `*_FILTERS` environment variables to run tests, benchmarks, and examples for specific packages.

For each of the commands, please run them from the root stdlib repository directory (not the package folder!).

You can use pattern matching to target specific packages as follows:

```bash
# Run tests for all packages in the math namespace:
make test TESTS_FILTER=".*/@stdlib/math/.*"

# Run benchmarks for specific packages (using OR pattern):
make benchmark BENCHMARKS_FILTER=".*/@stdlib/math/base/special/sin/.*|.*/@stdlib/math/base/special/cos/.*"
```

## Changed Packages

-   `@stdlib/math/base/special/tan`
-   `@stdlib/math/base/special/tand`
-   `@stdlib/math/base/special/tanh`

## Documentation Links

-   [make rules for running examples][make-docs-examples]
-   [make rules for running project unit tests][make-docs-test]
-   [make rules for running benchmarks][make-docs-benchmark]

[make-docs-examples]: https://github.com/stdlib-js/stdlib/blob/develop/tools/make/lib/examples/README.md
[make-docs-test]: https://github.com/stdlib-js/stdlib/blob/develop/tools/make/lib/test/README.md
[make-docs-benchmark]: https://github.com/stdlib-js/stdlib/blob/develop/tools/make/lib/benchmark/README.md"

Planeshifter committed:
build: also check for capital X when verifying contributing guidelines acknowledgment

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: passed
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter commented:
/stdlib rebase

kgryte commented:
@Planeshifter Is there a reason you keep force pushing to branches? This ends up messing with people's (i.e., my) local copies of your PRs.

kgryte commented:
Also, `s/*_FILTERS/*_FILTER`. The filter environment variables don't end with an `S`.

kgryte commented:
As an additional point, your OR example can be simplified:

```
make benchmark BENCHMARKS_FILTER=".*/@stdlib/math/base/special/(sin|cos)/.*"
```

Planeshifter commented:
@kgryte  I accidentally pushed a commit to this PR that was supposed to go to `develop` and is unrelated to this PR. I meant to fix this via a force push, but somehow the commands I would usually do to accomplish this (i.e., `git reset HEAD~1 --hard` followed by `git push origin philipp/package-make-commands-comment --force`) didn't work. There is definitely something strange going on with our `git` commit hashes;  potentially due to the pre-commit hooks? We should look into this; similar to @rreusser, I have recently encountered some strange behavior with pushes and pulls.

kgryte commented:
Yeah, we may need to abandon appending commit meta data to commits.

Planeshifter committed:
build: address PR feedback

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: passed
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter commented:
@kgryte Only the `pre-push` report does invoke `git commit --amend`. My hunch is that's the root of the problem because when we amend a commit, we are essentially creating a new commit with a different hash. The `pre-push` results seem less important than the `pre-commit` ones, so maybe we could consider disabling these to start with?

kgryte commented:
Yeah that is fine. Both are proxy for whether a dev has setup their local env. Feel free to change. I am still down and out.

kgryte commented on .github/workflows/scripts/package_commands_comment:
```diff
@@ -0,0 +1,253 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Script to post a comment with relevant `make` commands tailored to package changed in a PR.
```
```suggestion
# Script to post a comment with relevant `make` commands tailored to packages changed in a PR.
```

kgryte commented on .github/workflows/scripts/package_commands_comment:
```diff
@@ -0,0 +1,253 @@
+		if [[ "${#c_files[@]}" -eq 0 ]]; then
+			comment="Hello! 👋
+
+Pro-tip: Use the below \`make\` commands during local development to ensure that all tests, examples, and benchmark files in your PR run successfully.
```
```suggestion
Pro-tip: Use the \`make\` commands below during local development to ensure that all tests, examples, and benchmark files in your PR run successfully.
```

kgryte commented on .github/workflows/scripts/package_commands_comment:
```diff
@@ -0,0 +1,253 @@
+		else
+			comment="Hello! 👋
+
+Pro-tip: Use the below \`make\` commands during local development to ensure that all tests, examples, and benchmark files in your PR run successfully.
```
```suggestion
Pro-tip: Use the \`make\` below commands during local development to ensure that all tests, examples, and benchmark files in your PR run successfully.
```

kgryte commented on .github/workflows/scripts/package_commands_comment:
```diff
@@ -0,0 +1,253 @@
+
+For each of the commands, please run them from the root stdlib repository directory (not the package folder!).
+
+To build the native add-on,
```
```suggestion
To build a native add-on,
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on .github/workflows/scripts/package_commands_comment:
```diff
@@ -0,0 +1,253 @@
+		on_error 1
+	fi
+
+	# Fetch changed files in pull request (no need to paginate because skipping PRs affecting multiple packages anyway):
```
```suggestion
	# Fetch changed files in pull request:
```

kgryte committed:
docs: update comment

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on .github/workflows/scripts/package_commands_comment:
```diff
@@ -0,0 +1,253 @@
+## Documentation Links
+
+-   [make rules for running examples][make-docs-examples]
+-   [make rules for running project unit tests][make-docs-test]
```
```suggestion
-   [make rules for running unit tests][make-docs-test]
```

kgryte committed:
chore: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM. Although, in general, I would really like to get away from embedding documentation directly into Bash scripts like this.

Ideally, we would create separate Markdown template files which can be linted and maintained just like other Markdown files. Having all these escapes and unorthodox indentation practices thwarts find-and-replace and general maintenance practices.
